### PR TITLE
[pem] Create webpush certificates

### PR DESCRIPTION
**Pull request by jhanggi**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
Spaceship has [just about everything](https://github.com/fastlane/fastlane/pull/6943) it needs to support provisioning Web Push certificates, but you couldn't provision them because `Spaceship::App.all` does not include Web Push identifiers.

All that is really needed to make this work is to search `Spaceship::WebsitePush` instead when you're trying to provision a `Certificate::WebsitePush`. 

This should take care of https://github.com/fastlane/fastlane/pull/6943

Example usage of the PEM changes: `fastlane pem --website_push --app_identifier web.com.example.testing_app`

The addition of `specialIdentifierDisplayId` in the `submitCertificateRequest.action` request _might_ help with this issue: https://github.com/fastlane/fastlane/issues/10200 It probably won't completely address it, but hopefully it helps make some progress.

This was tested using `fastlane pem` as well as within my app, which also uses Spaceship to provision the push ID.

I'm open to suggestions on how to make this better and support more of the certificate types in the future.

### CODE REVIEW

**Comment by jhanggi** [comment]
I signed it!

**Comment by janpio** [comment]
Hey @jhanggi, sorry that you didn't get _any_ reaction until now. 

I don't know anything about Webpush, is this for this?
https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/NotificationProgrammingGuideForWebsites/PushNotifications/PushNotifications.html#//apple_ref/doc/uid/TP40013225-CH3-SW1

It would be really nice if you could take care of the merge conflicts and merge onto current `master`. This might unfortunately be a bit of work as during that timeframe we changed our coding style guide so that all method calls use parentheses - hope this is not too much work.

**Comment by jhanggi** [comment]
That's exactly what it's for. It's for sending push notifications to Safari. 

Most of the portal pieces were already there in `spaceship/lib/spaceship/portal/webpush.rb`. You could already create the webpush id, just not the certificate. For the cert, the cli options and the fact that the parameter for the id is called `specialIdentifierId` rather than `apiIdId`  were the main things missing.

I've merged in from latest `master`. I could rebase if you'd prefer (this PR is so old, the actual commits might get lost in distant history). 

The only significant conflict was the move of the extra namespace in `Spaceship::App` to `Spaceship::Portal::App` (and then `Spaceship::WebsitePush` to `Spaceship::Portal::WebsitePush`.

Thanks for finally following up!

File: pem/lib/pem/options.rb [hunk 1]
```diff
@@ -15,6 +15,11 @@ def self.available_options
                                      description: "Renew the development push certificate instead of the production one",
                                      is_string: false,
                                      default_value: false),
+        FastlaneCore::ConfigItem.new(key: :website_push,
```
Comment (janpio) [inline pem/lib/pem/options.rb]: this new param conflicts with `development`, right? I think there is a way to note that.

File: spaceship/lib/spaceship/portal/certificate.rb [hunk 2]
```diff
@@ -289,7 +294,7 @@ def create!(csr: nil, bundle_id: nil)
 
           # look up the app_id by the bundle_id
           if bundle_id
-            app = Spaceship::Portal::App.set_client(client).find(bundle_id)
+            app = portal_type.set_client(client).find(bundle_id)
```
Comment (janpio) [inline spaceship/lib/spaceship/portal/certificate.rb]: maybe I missed it in the code, but where does this come from when it is not a website_push? For that case you added the definition above in line 133 - but for production or development?

File: spaceship/lib/spaceship/portal/certificate.rb [hunk 3]
```diff
@@ -303,6 +308,12 @@ def create!(csr: nil, bundle_id: nil)
           response['certificateTypeDisplayId'] = response['certificateType']['certificateTypeDisplayId']
           self.new(response)
         end
+
+        # Default portal class to use when finding by bundle_id
```
Comment (janpio) [inline spaceship/lib/spaceship/portal/certificate.rb]: maybe move this up so the definition is before the first usage of it?
(honestly don't know if this matters for the actual code, but I just wondered where this came from in line 297, wrote a comment, and then just found this here when scrolling down later... so having it up would have a benefit)

File: spaceship/spec/portal/portal_stubbing.rb [hunk 4]
```diff
@@ -192,7 +192,7 @@ def adp_stub_certificates
       stub_request(:get, "https://developer.apple.com/services-account/QH65B2/account/ios/certificate/downloadCertificateContent.action?certificateId=XC5PH8DAAA&teamId=XXXXXXXXXX&type=R58UK2EAAA").
         to_return(status: 200, body: adp_read_binary_fixture_file('aps_development.cer')) # note: binary!
       stub_request(:post, "https://developer.apple.com/services-account/QH65B2/account/ios/certificate/submitCertificateRequest.action").
-        with(body: { "appIdId" => "2HNR359G63", "csrContent" => adp_read_fixture_file('certificateSigningRequest.certSigningRequest'), "type" => "BKLRAVXMGM", "teamId" => "XXXXXXXXXX" }).
```
Comment (janpio) [inline spaceship/spec/portal/portal_stubbing.rb]: is there no value in keeping the old test as well?

**Review by janpio — COMMENTED** [review]
Nah, merging was just fine - same result. (We will squash and merge later to create a new commit with the changes.)

I added a few comments trying to understand how this works.

I will also try to get someone else to have a look at the code for a review.

How confident are you in the code in the current state? Tested locally, also the functionality it touches (production and development push certificates)?

**Comment by joshdholtz** [comment]
Would like some 👀 on my two latest commits 😊 I tested this with both web push, dev, and dist certs and things were 👌

**Review by getaaron — APPROVED** [review]

**Review by janpio — APPROVED** [review]
Looks good!

**Comment by janpio** [comment]
There is a report that this is causing issues in the Swift variation of Fastlane: https://github.com/fastlane/fastlane/issues/14427

**Pull request merged**